plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IncrementalInMemoryRecordCursor.java

```diff
@@ -0,0 +1,199 @@
+
+public class IncrementalInMemoryRecordCursor
+        implements RecordCursor
+{
+    private final List<Type> types;
+    private CloseableIterable<List<Object>> records;
+    private CloseableIterator<List<Object>> iterator;
+    private List<Object> record;
+    private long completedBytes;
+    private boolean closed;
+
+    public IncrementalInMemoryRecordCursor(List<Type> types, CloseableIterable<List<Object>> records)
+    {
+        this.types = types;
+        this.records = records;
+        this.iterator = records.iterator();
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return completedBytes;
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return 0;
+    }
+
+    @Override
+    public Type getType(int field)
+    {
+        return types.get(field);
+    }
+
+    @Override
+    public boolean advanceNextPosition()
+    {
+        if (closed || !iterator.hasNext()) {
+            record = null;
+            return false;
+        }
+        record = iterator.next();
+        completedBytes += sizeOf(record);
+
+        return true;
+    }
+
+    @Override
+    public boolean getBoolean(int field)
+    {
+        checkArgument(record != null, "no current record");
+        requireNonNull(record.get(field), "value is null");
+        return (Boolean) record.get(field);
+    }
+
+    @Override
+    public long getLong(int field)
+    {
+        checkArgument(record != null, "no current record");
+        requireNonNull(record.get(field), "value is null");
+        return ((Number) record.get(field)).longValue();
+    }
+
+    @Override
+    public double getDouble(int field)
+    {
+        checkArgument(record != null, "no current record");
+        requireNonNull(record.get(field), "value is null");
+        return (Double) record.get(field);
+    }
+
+    @Override
+    public Slice getSlice(int field)
+    {
+        checkArgument(record != null, "no current record");
+        Object value = record.get(field);
+        requireNonNull(value, "value is null");
+        if (value instanceof byte[]) {
+            return Slices.wrappedBuffer((byte[]) value);
+        }
+        if (value instanceof String) {
+            return Slices.utf8Slice((String) value);
+        }
+        if (value instanceof Slice) {
+            return (Slice) value;
+        }
+        throw new IllegalArgumentException("Field " + field + " is not a String, but is a " + value.getClass().getName());
+    }
+
+    @Override
+    public Object getObject(int field)
+    {
+        checkArgument(record != null, "no current record");
+        Object value = record.get(field);
+        requireNonNull(value, "value is null");
+        return value;
+    }
+
+    @Override
+    public boolean isNull(int field)
+    {
+        checkArgument(record != null, "no current record");
+        return record.get(field) == null;
+    }
+
+    @Override
+    public void close()
```

Comment:
except `close()` are there major differences between this class and the methods from `io.trino.spi.connector.InMemoryRecordSet` ? Would it make sense to rather extend the before mentioned class instead writing a new one?

Reply:
Yes, only the `close()` method was additional. I have renamed it with `CloseableInMemoryRecordCursor` which will extend the `InMemoryRecordSet.InMemoryRecordCursor` and implement the `close()` method.

Reply:
Thanks for pointing it out. The code is much reduced now.